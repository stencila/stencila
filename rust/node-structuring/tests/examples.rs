use std::{
    fs::{read_dir, read_to_string},
    path::PathBuf,
    str::FromStr,
};

use eyre::{OptionExt, Result, bail};
use insta::assert_json_snapshot;
use itertools::Itertools;

use stencila_codec::{Codec, StructuringOperation, StructuringOptions, stencila_format::Format};
use stencila_codec_json::JsonCodec;
use stencila_codec_markdown::MarkdownCodec;
use stencila_node_structuring::structuring;
use stencila_schema::{Node, Primitive};

/// Structure example Markdown or JSON files using the structuring operation
/// specified in the first three parts of their name, or multiple operations for
/// files with numeric prefixes.
///
/// Run using:
///
/// cargo insta test -p stencila-node-structuring
///
/// and then review snapshots with:
///
/// cargo insta review
#[tokio::test]
async fn examples() -> Result<()> {
    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/examples");

    for entry in read_dir(&dir)?.flatten() {
        let path = entry.path();
        let format = Format::from_path(&path);

        let content = read_to_string(&path)?;

        let (mut node, ..) = match format {
            Format::Markdown => MarkdownCodec.from_str(&content, None).await?,
            Format::Json => JsonCodec.from_str(&content, None).await?,
            _ => bail!("Unsupported formatted"),
        };

        let id = path
            .file_stem()
            .map(|name| name.to_string_lossy().to_string())
            .ok_or_eyre("No file stem")?;

        let Node::Article(article) = &node else {
            bail!("Expected an article")
        };

        let operations = if let Some(Primitive::Array(ops)) = article
            .options
            .extra
            .as_ref()
            .and_then(|extra| extra.get("structuring").cloned())
        {
            ops.iter()
                .filter_map(|op| match op {
                    Primitive::String(op) => StructuringOperation::from_str(op).ok(),
                    _ => None,
                })
                .collect_vec()
        } else {
            // Fall back to operation from filename for single operation tests
            let op = id.split("-").take(3).collect_vec().join("-");
            if let Ok(operation) = StructuringOperation::from_str(&op) {
                vec![operation]
            } else {
                bail!("Unable to determine structuring operations for {}", path.display());
            }
        };

        structuring(&mut node, StructuringOptions::new(operations, []));

        assert_json_snapshot!(id, node);
    }

    Ok(())
}
