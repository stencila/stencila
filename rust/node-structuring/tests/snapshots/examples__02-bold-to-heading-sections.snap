---
source: rust/node-structuring/tests/examples.rs
expression: node
---
{
  "type": "Article",
  "frontmatter": "structuring:\n- paragraphs-to-headings\n- headings-to-sections",
  "content": [
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "This tests that bold paragraphs like "
          }
        },
        {
          "type": "Strong",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "Methods"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " are first converted to headings with appropriate level (level 1 for primary sections), then wrapped in section elements. This interaction is important because the heading level determination affects section hierarchy and primary sections must be recognized correctly."
          }
        }
      ]
    },
    {
      "type": "Section",
      "sectionType": "Introduction",
      "content": [
        {
          "type": "Heading",
          "level": 1,
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "Introduction"
              }
            }
          ]
        },
        {
          "type": "Paragraph",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "The bold paragraph above should be converted to a level 1 heading because \"Introduction\" is a primary section type, then wrapped in a section element."
              }
            }
          ]
        }
      ]
    },
    {
      "type": "Section",
      "sectionType": "Methods",
      "content": [
        {
          "type": "Heading",
          "level": 1,
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "Methods"
              }
            }
          ]
        },
        {
          "type": "Paragraph",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "This bold paragraph should also become a level 1 heading since \"Methods\" is a primary section type, then be wrapped in a section."
              }
            }
          ]
        },
        {
          "type": "Section",
          "sectionType": "StatisticalAnalysis",
          "content": [
            {
              "type": "Heading",
              "level": 2,
              "content": [
                {
                  "type": "Text",
                  "value": {
                    "string": "Data Analysis"
                  }
                }
              ]
            },
            {
              "type": "Paragraph",
              "content": [
                {
                  "type": "Text",
                  "value": {
                    "string": "This bold paragraph should become a heading (level 2 since it's not a primary section type) and then be wrapped appropriately in the section hierarchy."
                  }
                }
              ]
            }
          ]
        }
      ]
    }
  ],
  "structuring": [
    "paragraphs-to-headings",
    "headings-to-sections"
  ]
}
